### PR TITLE
Product category attach

### PR DESCRIPTION
**Pull request by jayesh-parmar**

### CODE REVIEW

File: app/Http/Controllers/ProductController.php [hunk 1]
```diff
@@ -13,36 +14,42 @@ class ProductController extends Controller
...
+        $sizes = Size::select('id', 'name')->get();
+        $categories = Category::select('id', 'name')->get();
+
+        return view('admin.pages.product.form', compact('brands', 'colors', 'sizes', 'categories'));
     }
     public function store(ProductValidation $request)
     {
-        Product::create($request->validated());
+        $product = Product::create($request->validated());
+        $product->attachCategories($request->input('category_ids'));
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: `attachCategories` method move to here(controller) and it should be private function.

File: app/Http/Controllers/ProductController.php [hunk 2]
```diff
@@ -13,36 +14,42 @@ class ProductController extends Controller
...
         $brands = Brand::select('id', 'name')->get();
         $colors = Color::select('id', 'name')->get();
-        $sizes = Size::select('id', 'name')->get(); 
+        $sizes = Size::select('id', 'name')->get();
+        $categories = Category::select('id', 'name')->get();
+
         $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->find($productId);
 
-        return view('admin.pages.product.form', compact('product', 'brands', 'colors', 'sizes'));
+        return view('admin.pages.product.form', compact('product', 'brands', 'colors', 'sizes', 'categories'));
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: categories attached with product are automatic selected?

File: app/Http/Controllers/ProductController.php [hunk 3]
```diff
@@ -13,36 +14,42 @@ class ProductController extends Controller
...
     }
 
     public function update(ProductValidation $request, Product $product)
     {
         $product->update($request->validated());
+        $product->categories()->sync($request->input('category_ids', []));
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: never use $request->input() value. you have $request->validated() array.

File: app/Models/Category.php [hunk 4]
```diff
@@ -19,4 +19,8 @@ public function children()
     {
         return $this->hasMany(Category::class, 'parent_id');
     }
+    public function products()
+    {
+        return $this->belongsToMany(Product::class);
+    }
```
Comment (rajan-parmar) [inline app/Models/Category.php]: required?

File: app/Http/Controllers/ProductController.php [hunk 5]
```diff
@@ -13,36 +14,42 @@ class ProductController extends Controller
...
+
         $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->find($productId);
 
-        return view('admin.pages.product.form', compact('product', 'brands', 'colors', 'sizes'));
+        return view('admin.pages.product.form', compact('product', 'brands', 'colors', 'sizes', 'categories'));
     }
 
     public function update(ProductValidation $request, Product $product)
     {
         $product->update($request->validated());
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php:59]: now you are attaching category with product. so what if category trying to delete and that is connected with product?

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/ProductController.php (hunk 5 above)
Comment (jayesh-parmar) [reply app/Http/Controllers/ProductController.php:50]: sir error ma samjatu nathi

File: app/Http/Controllers/ProductController.php [hunk 6]
```diff
@@ -33,22 +33,35 @@ public function store(ProductValidation $request)
...
 
-        $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->find($productId);
+       $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->with('categories')->find($productId);
+       $productCategories = $product->categories->pluck('id')->toArray();
+
+       foreach ($productCategories as $categoryId) {
+        if (!$categories->contains('id', $categoryId)) {
+                $product->categories()->detach($categoryId);
+            }
+        }
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: any reason?

File: app/Http/Controllers/ProductController.php [hunk 7]
```diff
@@ -33,22 +33,35 @@ public function store(ProductValidation $request)
...
+
     public function edit(string $productId)
     {
         $brands = Brand::select('id', 'name')->get();
         $colors = Color::select('id', 'name')->get();
         $sizes = Size::select('id', 'name')->get();
         $categories = Category::select('id', 'name')->get();
 
-        $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->find($productId);
+       $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->with('categories')->find($productId);
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: always load the required columns while calling the relation

File: app/Http/Requests/ProductValidation.php [hunk 8]
```diff
@@ -15,6 +15,7 @@ public function rules(): array
             'color_id' => ['nullable', 'string', 'exists:colors,id'],
             'size_id' => ['nullable', 'string', 'exists:sizes,id'],
             'status' => ['boolean'],
+            'category_ids' => ['array'],
```
Comment (rajan-parmar) [inline app/Http/Requests/ProductValidation.php]: required.
also validate the array nested value

File: app/Http/Controllers/ProductController.php [hunk 9]
```diff
@@ -33,22 +33,35 @@ public function store(ProductValidation $request)
...
+        }
 
         return view('admin.pages.product.form', compact('product', 'brands', 'colors', 'sizes', 'categories'));
     }
 
     public function update(ProductValidation $request, Product $product)
     {
         $product->update($request->validated());
-        $product->categories()->sync($request->input('category_ids', []));
+        $product->categories()->sync($request->validated(['category_ids'] ?? []));
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]:
```suggestion
        $product->categories()->sync($request->validated(['category_ids']));
```

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/ProductController.php [hunk 10]
```diff
@@ -46,22 +48,16 @@ public function edit(string $productId)
         $sizes = Size::select('id', 'name')->get();
         $categories = Category::select('id', 'name')->get();
 
-       $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->with('categories')->find($productId);
+       $product = Product::select('id', 'name', 'brand_id', 'size_id', 'color_id', 'status')->with('categories:id,name')->find($productId);
        $productCategories = $product->categories->pluck('id')->toArray();
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: $productCategories anywhere use?

File: app/Http/Requests/ProductValidation.php [hunk 11]
```diff
@@ -15,7 +15,8 @@ public function rules(): array
             'color_id' => ['nullable', 'string', 'exists:colors,id'],
             'size_id' => ['nullable', 'string', 'exists:sizes,id'],
             'status' => ['boolean'],
-            'category_ids' => ['array'],
+            'category_ids' => ['required','array'],
+            'category_ids.*' => ['exists:categories,id'],
```
Comment (rajan-parmar) [inline app/Http/Requests/ProductValidation.php]:
```suggestion
            'category_ids.*' => ['string', 'exists:categories,id'],
```

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/CategoryController.php [hunk 12]
```diff
@@ -42,8 +42,10 @@ public function destroy(string $categoryId)
         if (Category::where('parent_id', $categoryId)->exists()) {
             return redirect()->route('categories.index')->with('error', 'This category cannot be deleted as there are one or more child categories attached.');
         }
-        Category::where('id', $categoryId)->delete();
         
+        if(Category::where('id', $categoryId))
+        return redirect()->route('categories.index')->with('error', 'This category cannot be deleted as there are one or more product attached.');
```
Comment (rajan-parmar) [inline app/Http/Controllers/CategoryController.php]: follow the sequence
if () {
}

File: app/Http/Requests/ProductValidation.php [hunk 13]
```diff
@@ -15,6 +15,7 @@ public function rules(): array
             'color_id' => ['nullable', 'string', 'exists:colors,id'],
             'size_id' => ['nullable', 'string', 'exists:sizes,id'],
             'status' => ['boolean'],
+            'category_ids.*' => ['string', 'exists:categories,id'],
```
Comment (rajan-parmar) [inline app/Http/Requests/ProductValidation.php:19]: why remove the previous changes?

File: app/Models/Product.php [hunk 14]
```diff
@@ -30,4 +30,12 @@ public function size()
...
         return $this->belongsTo(Size::class);
     }
+    public function categories()
+    {
+        return $this->belongsToMany(Category::class);
+    }
+    public function attachCategories($categoryIds)
+    {
+        $this->categories()->attach($categoryIds);
+    }
```
Comment (rajan-parmar) [inline app/Models/Product.php]: not required

File: app/Http/Controllers/ProductController.php [hunk 15]
```diff
@@ -13,36 +14,50 @@ class ProductController extends Controller
...
+        $categories = Category::select('id', 'name')->get();
+
+        return view('admin.pages.product.form', compact('brands', 'colors', 'sizes', 'categories'));
+    }  
+
     public function store(ProductValidation $request)
     {
-        Product::create($request->validated());
+        $product = Product::create($request->validated());
+        $product->attachCategories($request->validated(['category_ids']));
```
Comment (rajan-parmar) [inline app/Http/Controllers/ProductController.php]: create variable for $request->validated()
why use again attachCategories from the product? I already said to make private method see link:
https://github.com/jayesh-parmar/inventory-management/pull/25#discussion_r1342357701

**Review by rajan-parmar — CHANGES_REQUESTED** [review]

**Review by rajan-parmar — APPROVED** [review]

**Pull request merged**